File: mayonez-base/src/main/java/mayonez/input/MappingStatus.java
package mayonez.input;

import mayonez.annotations.*;

/**
 * Represents the status for a key or button, or whether it is pressed,
 * held, or released.
 *
 * @author SlavSquatSuperstar
 */
@ExperimentalFeature
class MappingStatus {

    /**
     * If the input listener detected this mapping to be down.
     */
    private boolean down;

    /**
     * If this mapping was newly pressed in this frame.
     */
    private boolean pressed;

    /**
     * If this mapping is continuously held down.
     */
    private boolean held;

    MappingStatus() {
        down = false;
        pressed = false;
        held = false;
    }

    boolean isDown() {
        return down;
    }

    void setDown(boolean down) {
        this.down = down;
    }

    boolean isPressed() {
        return pressed;
    }

    void setPressed() {
        pressed = true;
    }

    public boolean isHeld() {
        return held;
    }

    void setHeld() {
        pressed = false;
        held = true;
    }

    boolean isReleased() {
        return !pressed && !held;
    }

    void setReleased() {
        pressed = false;
        held = false;
    }

}

File: mayonez-base/src/main/kotlin/mayonez/input/MappingStatus.kt
package mayonez.input

/**
 * Tracks the status for a key or button across multiple frames.
 *
 * @author SlavSquatSuperstar
 */
internal class MappingStatus {

    /** If the input listener detected this mapping to be down. */
    var down: Boolean = false
    private var state: InputState = InputState.RELEASED

    // Getters and Setters

    val pressed: Boolean
        get() = state == InputState.PRESSED

    val held: Boolean
        get() = state == InputState.HELD

    val released: Boolean
        get() = state == InputState.RELEASED

    fun setReleased() {
        if (!released) state = InputState.RELEASED
    }

    fun updateIfDown() {
        if (released) {
            state = InputState.PRESSED
        } else if (pressed) {
            state = InputState.HELD
        }
    }

}